### ISSUE TITLE
 asterisk-chan-dongle: open('/var/lock/LCK..ttyUSB*') failed: Permission denied

### ISSUE THREAD

**Issue by 532910** [openwrt/telephony#685]
Maintainer: @jslachta
Environment: cortexa53 friendlyarm_nanopi-neo2 21.02.0

Description:
1. Starting with 21.02.0 asterisk runs as `asterisk` user.
2. chowning `/dev/ttyUSB*` to `asterisk:dialout` manually gives another error:

```
ERROR[4594]: chan_dongle.c:140 lock_create: open('/var/lock/LCK..ttyUSB2') failed: Permission denied
WARNING[4594]: chan_dongle.c:225 opentty: unable to open /dev/ttyUSB2: Resource busy
```

**Comment by micmac1** [comment]
I doubt /var/lock even exists. You might have to create it.

OpenWrt doesn't have a mechanism to set device permissions (think "udev" like) as far as I'm aware. So in my opinion this is not a problem of the asterisk package. What people do to get by is update the init script and set the permissions (or create directories like /var/lock/xyz) before the application is launched.

**Comment by micmac1** [comment]
Actually you may be able to use OpenWrt hotplug for that [link](https://openwrt.org/docs/guide-user/base-system/hotplug).

**Comment by 532910** [comment]
I tried, but it doesn't provide device name (like /dev/ttyUSB0)

**Comment by micmac1** [comment]
Provides other stuff that looks usable though, according to the documentation I linked. Like VENDOR and whatnot.

**Comment by 532910** [comment]
There is no VENDOR described on [hotplug](https://openwrt.org/docs/guide-user/base-system/hotplug) page.
I tried [troubleshooting](https://openwrt.org/docs/guide-user/base-system/hotplug#usagetroubleshooting) script and found no way to set /dev/ttyUSB0 permission.

**Comment by micmac1** [comment]
It's in the PRODUCT var

**Comment by rojer** [comment]
`/var/lock` does exist, actually, and is populated with other stuff:
```
root@SP-BR1:~# ls -l /var/lock
-rw-r--r--    1 root     root             0 Oct  1 19:59 procd_asterisk.lock
-rw-r--r--    1 root     root             0 Oct  1 19:55 procd_cron.lock
-rw-r--r--    1 root     root             0 Oct  1 19:59 procd_dnsmasq.lock
-rw-r--r--    1 root     root             0 Oct  1 19:55 procd_dropbear.lock
-rw-r--r--    1 root     root             0 Oct  1 19:55 procd_firewall.lock
-rw-r--r--    1 root     root             0 Oct  1 19:55 procd_gpio_switch.lock
...
```

problem is, default directory permissions don't allow other files in this directory to be created:

```
root@SP-BR1:~# ls -lad /var/lock/                                                                                                                  
drwxr-xr-x    2 root     root           420 Oct  1 19:59 /var/lock/
```

i think it should have permissions similar to /tmp, i.e. `1777` - to allow anyone create new files but not mess with other people's files

**Comment by rojer** [comment]
fwiw, solved with the following hotplug script:

```
root@SP-BR1:~# cat /etc/hotplug.d/usb/30-ttyUSB
chmod 1777 /var/lock
chown asterisk /dev/ttyUSB*
```

@micmac1 does it make sense to add it to the package?

**Comment by micmac1** [comment]
I don't think so. There may be some /dev/ttyUSB* devices that users don't want to have handed over to asterisk. I think it's good that you shared what you have done, so others can find it.

Have a good weekend!
Seb

**Comment by rojer** [comment]
alternatively, we could auto-add asterisk to group dialout that has rw access to ttyUSB devices.

**Comment by micmac1** [comment]
I don't think that works. I think there is an issue that was opened here where somebody reported it doesn't, already a while ago.

**Comment by micmac1** [comment]
Maybe I'm confusing this with /dev/vmmc* used by chan-lantiq, though.

If adding asterisk user to dialout group works for you then that's good, of course. I just don't think this should be done by default, by the package. I'd like the default user to have the least amount of rights possible. I'd rather somebody added a bit of info around this topic to the wiki and mention the possibility (if it works).

Kind regards,
Seb

>I don't think that works. I think there is an issue that was opened here where somebody reported it doesn't, already a while ago.

**Comment by dhewg** [comment]
As for the lock directory, that whole locking dance looks fishy... but as a first step you could use the configured astrundir, that's what it's for and asterisk itself sets it up with the right permissions.

Try:
`#include <asterisk/path.h>`
and replace this #if block: 
https://github.com/wdoekes/asterisk-chan-dongle/blob/master/chan_dongle.c#L121
with `return snprintf(buf, length, "%s/LCK..%s", ast_config_AST_RUN_DIR, basename);`

**Comment by rojer** [comment]
indeed, simply adding asterisk to dialout does not work.
i've looked into why and the reason is the way privileges are dropped when starting the asterisk process.
currently this is done by passing `-U asterisk -G asterisk` to the command, so it starts as root and the drops to `asterisk` on its own. while doing that, it does not apply ancillary groups as evidenced by checking runtime group membership of the running process:

```
root@SP-BR1:~# cat /proc/3557/status                                                                                                               
Name:   asterisk                                                                                                                                   
...
Uid:    385     385     385     385                                      
Gid:    385     385     385     385                           
FDSize: 32                                                   
Groups:                                                                  
```
note the empty `Groups` list.
however, if instead we use `procd`'s `user` param instead, it performs the user switch properly and the asterisk process comes up with dialout group membership:

```
root@SP-BR1:~# cat /proc/7331/status
Name:   asterisk
...
Uid:    385     385     385     385
Gid:    385     385     385     385
FDSize: 32
Groups: 20 385 385 
```

20 is the dialout group.

so, the following small patch to `/etc/init.d/asterisk` did the trick for me:
```
# diff -u asterisk.bak asterisk
--- asterisk.bak        2021-10-01 22:00:45.913387099 +0100
+++ asterisk    2021-10-01 22:00:45.913387099 +0100
@@ -49,10 +49,9 @@
   done
 
   procd_open_instance
+  procd_set_param user $NAME
   procd_set_param command $COMMAND
   procd_append_param command \
-    -G "$NAME" \
-    -U "$NAME" \
     $options \
     -f
   # forward stderr to logd
```

**Comment by dhewg** [comment]
Removing -U and/or -G will likely break setting up the astvarrun dir.

Try
```
/etc/init.d/asterisk stop
rm -rf /var/run/asterisk
/etc/init.d/asterisk start
```
with your procd patch. I think it shoud fail creating /var/run/asterisk then?

**Comment by micmac1** [comment]
That's one side of the medallion. The other would be asterisk setting some things up initially, like realtime priority or that sort of thing (not saying that would work with musl).

**Comment by dhewg** [comment]
Yeah, I don't think dropping -U/-G is the right way.

Adding dialout to asterisk likely doesn't work because of this: https://github.com/asterisk/asterisk/blob/master/main/asterisk.c#L3793
But removing that is maybe not such a good idea too ;)

You probably just need to extend your hotplug script to check for specific usb vid/pid s, see
https://openwrt.org/docs/guide-user/base-system/hotplug#restart_wlan_on_usb_wifi_hotplug

**Comment by rojer** [comment]
`dialout` membership is the right way to get access to modem-like devices, so in that sense `dialout` is a _needed_ group and should not be dropped (i am referring to the message in asterisk.c that accompanies `setgroups`).
what if `-G` is extended to take multiple groups instead of one value? could be smth like `-G asterisk -G dialout` or `-G asterisk,dialout`.

**Comment by dhewg** [comment]
That might work too, sure.

I didn't check at all if this works out, but is there an actually need for the asterisk group? If the only reason for it is to drop privileges, can we just use -U asterisk -G dialout?

**Comment by micmac1** [comment]
I got curious and checked what happens with freeswitch. There the init script also provides user and group. So I added the freeswitch user to the dialout group:

```
id freeswitch
uid=372(freeswitch) gid=372(freeswitch) groups=372(freeswitch),20(dialout),372(freeswitch)
```

Then I start freeswitch and get this:

```
Uid:    372     372     372     372
Gid:    372     372     372     372
FDSize: 32
Groups:
```

So the same that you found with asterisk I reckon.

Then I removed only the line 

`-g "$NAME"`

from the init script and restarted freeswitch, et voila:

```
TracerPid:	0
Uid:	372	372	372	372
Gid:	372	372	372	372
FDSize:	32
Groups:	20 372 372 
NStgid:	32409
```

So like this (only handing freeswitch the user) it seems we have the best of both worlds.

Then I installed asterisk. I start it and then:

```
TracerPid:	0
Uid:	385	385	385	385
Gid:	385	385	385	385
FDSize:	32
Groups:	 
NStgid:	1337
```

Then I remove the

`-G "$NAME"`

line from init script and restart. Now this:

```
TracerPid:	0
Uid:	385	385	385	385
Gid:	385	385	385	385
FDSize:	32
Groups:	20 385 385 
NStgid:	1720
```

So maybe all we need to do is remove the setting of the group.

I don't have any usb modem so can't confirm it actually works. Can you verify?

Regards,
Seb

**Comment by dhewg** [comment]
I checked debian earlier, and they too just use -U and not -G.

**Comment by rojer** [comment]
ok, great. so looks like we need to remove -G from asterisk init script and modify the cdongle package to add asterisk to dialout.

**Comment by micmac1** [comment]
Not sure about the dongle package adding the user to the group. I think the user should do that manually. Let it be the users choice. Just my 2 cents.

**Comment by rojer** [comment]
well, i'm all for choice but this is also a hurdle that everyone installing the package will have to overcome.
access to modem devices is provided through dialout group membership.
this package is about those. it's also not installed by default, so user did express desire to use the functionality by installing the package, meaning he'd like asterisk to have access to these devices.
so i think it makes sense, in the interest of UX.

coming from embedded development field, i can say that it's quite normal for packages to allow access to devices they are primarily intended for - normally that involves adding udev rules.

**Comment by micmac1** [comment]
You might be right. I'm just very hesitant when it comes to permissions.

**Comment by micmac1** [comment]
But maybe at least add a note in postinst that tells users "asterisk user has been added to dialout group". So everybody is aware and it's transparent.

**Comment by 532910** [comment]
It'd be nice if installing asterisk-chan-dongle package will add/start asterisk to the `dialout` group, but just a note about confiration this will be sufficient.

**Comment by 532910** [comment]
> It's in the PRODUCT var

```
# cat /etc/hotplug.d/usb/dongles 
logger -t hotplug $(env)
```

`logread -f`:
```
HOTPLUG_TYPE=usb
DEVPATH=/devices/platform/soc/1c1d000.usb/usb3/3-1/3-1.5/3-1.5:1.1
DEVICENAME=3-1.5:1.1
MODALIAS=usb:v12<cropped digit>
PRODUCT=12d1/1001/0
INTERFACE=255/255/255
```

@micmac1, yes, the `PRODUCT` var exist, but how it help to chown `/dev/ttuUSBXX`?

**Comment by rojer** [comment]
@532910 `basename $(echo /sys/bus/usb/devices/${DEVICENAME}/ttyUSB*)` should give you the name of the port.

**Comment by 532910** [comment]
Indeed, but it looks odd, perhaps it due to my attitude to templates.

**Comment by micmac1** [comment]
Backported fixes to 21.02. Closing issue.

Thanks!
